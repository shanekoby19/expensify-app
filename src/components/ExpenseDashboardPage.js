import React from 'react';
import ConnectedExpenseList from './ExpenseList';
import ConnectedExpenseListFilters from './ExpenseListFilters';
import ConnectedExpensesSummary from './ExpensesSummary'

export default () => (
    <div>
        <ConnectedExpensesSummary />
        <ConnectedExpenseListFilters />
        <ConnectedExpenseList />
    </div>
);