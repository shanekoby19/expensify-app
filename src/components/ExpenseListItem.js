import React from 'react';
import { Link } from 'react-router-dom';
import numeral from 'numeral';

export default ({ description, amount, createdAt, id }) => (
    <div>
        <h2>
            <Link to={`/edit/${id}`}>{description}</Link>
        </h2>
        <p>
            { numeral(amount / 100).format('$0,0.00') } 
             - 
            { createdAt.format('MMMM Mo') }
        </p>
    </div>
)