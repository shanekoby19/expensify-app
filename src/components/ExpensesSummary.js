import numeral from 'numeral';
import React from 'react';
import { connect } from 'react-redux';
import getAllExpenses from '../selectors/expenses';
import getTotalAmount from '../selectors/expenses-total'

export const ExpensesSummary = (props) => {
    const filteredExpenses = getAllExpenses(props.expenses, props.filters);
    return (
        <div>
            <p>Viewing {filteredExpenses.length} {filteredExpenses.length === 1 ? 'expense' : 'expenses'} totaling {numeral(getTotalAmount(filteredExpenses) / 100).format('$0,0.00')}.</p>
        </div>
    )
}

const mapStateToProps = (state) => ({
    expenses: state.expenses,
    filters: state.filters,
})

export default connect(mapStateToProps)(ExpensesSummary)