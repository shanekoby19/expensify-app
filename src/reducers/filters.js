import moment from 'moment'
import { v4 as uuidv4 } from 'uuid'

// Filters Reducer
const filtersReduceDefaultState = {
    text: ``,
    sortBy: 'date',
    startDate: moment().startOf('month'),
    startDateId: uuidv4(),
    endDate: moment().endOf('month'),
    endDateId: uuidv4(),
}

export default (state = filtersReduceDefaultState, action) => {
    switch(action.type) {
        case 'SET_TEXT_FILTER':
            return {
                ...state,
                text: action.text,
            }
        case 'SORT_BY_DATE':
            return {
                ...state,
                sortBy: 'date',
            }
        case 'SORT_BY_AMOUNT':
            return {
                ...state,
                sortBy: 'amount',
            }
        case 'SET_START_DATE':
            return {
                ...state,
                startDate: action.startDate,
            }
        case 'SET_END_DATE':
            return {
                ...state,
                endDate: action.endDate,
            }
        default:
            return state
    }
}