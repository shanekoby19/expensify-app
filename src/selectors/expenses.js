import moment from 'moment';

export default ( expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter(expense => {
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const textMatch = text ? expense.description.toLowerCase().includes(text.toLowerCase()) : true;
        
        return startDateMatch && endDateMatch && textMatch;
    }).sort((expenseOne, expenseTwo) => {
        if(sortBy === 'amount') {
            return expenseOne.amount >= expenseTwo.amount ? -1 : 1;
        }
        else if(sortBy === 'date') {
            return expenseOne.createdAt >= expenseTwo.createdAt ? -1 : 1;
        }
    })
}