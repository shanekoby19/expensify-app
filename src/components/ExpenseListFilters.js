import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates'

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null,
    };
    
    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }))
    }

    onTextChange = (e) =>  {
        e.target.value === 'date' ? this.props.sortByDate() : this.props.sortByAmount();
    }

    onSortChange = (e) =>  {
        this.props.setTextFilter(e.target.value);
    }

    render = () => {
        return (
            <div>
                <input 
                    type='text' 
                    value={this.props.filters.text} 
                    onChange={this.onSortChange}
                />
                <select 
                    value={this.props.filters.sortBy} 
                    onChange={this.onTextChange}
                >
                    <option value='date'>Date</option>
                    <option value='amount'>Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    startDateId={this.props.filters.startDateId}
                    endDate={this.props.filters.endDate}
                    endDateId={this.props.filters.endDateId}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    showClearDates={true}
                />
            </div>
        ) 
    }
}

const mapStateToProps = (state) => ({
        filters: state.filters,
})

const mapDispatchToProps = (dispatch => ({
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
}))

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);