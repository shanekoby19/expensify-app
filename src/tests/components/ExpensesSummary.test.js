import { test, expect } from "@jest/globals";
import React from 'react';
import { shallow } from 'enzyme'
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';
import { filters } from '../fixtures/filters';

test('should render a paragraph totaling a single expense', () => {
    const wrapper = shallow( <ExpensesSummary expenses={[expenses[0]]} filters={filters}/> );
    expect(wrapper).toMatchSnapshot();
})

test('should render a paragraph totaling all the expenses', () => {
    const wrapper = shallow(<ExpensesSummary expenses={expenses} filters={filters}/>);
    expect(wrapper).toMatchSnapshot();
})