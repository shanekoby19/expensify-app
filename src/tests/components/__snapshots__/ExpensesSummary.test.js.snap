// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a paragraph totaling a single expense 1`] = `
<div>
  <p>
    Viewing 
    1
     
    expense
     totaling 
    $1.95
    .
  </p>
</div>
`;

exports[`should render a paragraph totaling all the expenses 1`] = `
<div>
  <p>
    Viewing 
    3
     
    expenses
     totaling 
    $1,141.95
    .
  </p>
</div>
`;
